component/src/lib.rs

```
@@ -0,0 +1,127 @@
+
+    // The function for cleanup after the task has encountered an irrecoverable error
+    fn handle_irrecoverable(&self, irrecoverable: IrrecoverableError) -> Result<(), anyhow::Error>;
+}
+
+impl<M: Manageable + Send> Supervisor<M> {
+    /// Creates a new supervisor using a Manageable component.
+    pub fn new(component: M) -> Self {
+        let (_, tr_irrecoverable) = channel(10);
+        let (tx_cancellation, _) = oneshotChannel();
+        Supervisor {
+            irrecoverable_signal: tr_irrecoverable,
+            cancellation_signal: Some(tx_cancellation),
+            join_handle: None,
+            manageable: component,
+        }
+    }
+
+    /// Spawn calls the start function of the Manageable component and runs supervision.
+    pub async fn spawn(mut self) -> Result<(), anyhow::Error> {
+        let (tx_irrecoverable, tr_irrecoverable) = channel(10);
+        let (tx_cancellation, tr_cancellation) = oneshotChannel();
+
+        // this assumes the start_fn is populated
+        let wrapped_handle = self
+            .manageable
+            .start(tx_irrecoverable, tr_cancellation)
+            .await;
+
+        self.irrecoverable_signal = tr_irrecoverable;
+        self.cancellation_signal = Some(tx_cancellation);
+        self.join_handle = Some(wrapped_handle);
+
+        self.run().await
+    }
+
+    /// Run watches continuously for irrecoverable errors or JoinHandle completion.
+    pub async fn run(mut self) -> Result<(), anyhow::Error> {
+        // select statement that listens for the following cases:
+        //
+        // Irrecoverable signal incoming => log, terminate and restart
+        // completion of the task => already terminated, log and restart
+        //
+        // The handle_irrecoverable is run before the existing task gets
+        // cancelled by restart in the case that an irrecoverable signal
+        // was sent to us. This makes resource cleanup possible.
```

Probably, here if `self.join_handle.is_none()` we should return an error and not enter the loop, since we are trying to run the task without a proper channel? Or, if this is only intended to be ran by `spawn`, which is guaranteed to initialize the channels, maybe the function `run` should not be `pub` to avoid library user error in calling it before it is ready.

---

removing `pub`